test/Chainweb/Test/Cut.hs

```diff
@@ -142,92 +205,64 @@ testMine'
     -> Cut
     -> IO (Either MineFailure (T2 BlockHeader Cut))
 testMine' wdb n t payloadHash i c =
-    forM (createNewCut n (t c (_chainId i)) payloadHash i c) $ \p@(T2 h _) ->
-        p <$ insertWebBlockHeaderDb wdb h
-
--- | Block time generation that offsets from previous chain block in cut.
-offsetBlockTime :: TimeSpan Micros -> GenBlockTime
-offsetBlockTime offset cut cid = add offset t
-  where
-    BlockCreationTime t = cut ^?! ixg cid . blockCreationTime
-
-arbitraryBlockTimeOffset
-    :: TimeSpan Micros
-    -> TimeSpan Micros
-    -> T.Gen GenBlockTime
-arbitraryBlockTimeOffset lower upper = do
-    t <- genEnum (lower, upper)
-    return $ offsetBlockTime t
+    try (createNewCut (chainLookupM wdb) n (t c (_chainId i)) payloadHash i c) >>= \case
+        Right p@(T2 h _) -> Right p <$ insertWebBlockHeaderDb wdb h
+        e -> return e
 
 testMineWithPayloadHash
-    :: forall cid
+    :: forall cid hdb
     . HasChainId cid
-    => Nonce
+    => ChainValueCasLookup hdb BlockHeader
+    => hdb
+    -> Nonce
     -> Time Micros
     -> BlockPayloadHash
     -> cid
     -> Cut
     -> IO (Either MineFailure (T2 BlockHeader Cut))
-testMineWithPayloadHash n t payloadHash i c =
-    forM (createNewCut n t payloadHash i c) return
+testMineWithPayloadHash db n t ph cid c = try
+    $ createNewCut (chainLookupM db) n t ph cid c
 
 -- | Create a new block. Only produces a new cut but doesn't insert it into the
 -- chain database.
 --
+-- The creation time isn't checked.
+--
 createNewCut
     :: HasCallStack
+    => MonadCatch m
     => HasChainId cid
-    => Nonce
+    => (ChainValue BlockHash -> m BlockHeader)
+    -> Nonce
     -> Time Micros
     -> BlockPayloadHash
     -> cid
     -> Cut
-    -> Either MineFailure (T2 BlockHeader Cut)
-createNewCut n t pay i c = do
-    h <- note BadAdjacents $ newHeader . BlockHashRecord <$> newAdjHashes
-    unless (checkTarget (_blockTarget h) $ _blockPow h) $ Left BadNonce
-    c' <- first (\e -> error $ "Chainweb.Cut.createNewCut: " <> sshow e)
-        $ monotonicCutExtension c h
+    -> m (T2 BlockHeader Cut)
+createNewCut hdb n t pay i c = do
+    extension <- fromMaybeM BadAdjacents $ getCutExtension c i
```

Comment:
I must say `fromMaybeM` is a terrible name. `fromMaybe` doesn't cause/throw an error. Please change this to a more descriptive name. (It's sibling `???` is also pretty bad, and used like twice, so I would prefer we just ditch it).

Reply:
The name is consistent with the convention throughout the chainweb code base that functions that end with `M` handle failures within `MonadThrow` (I think the `M` was originally inspired by the `M` in `throwM`).

This function has been in the code base since the very beginning. I think changing it is beyond the scope of this PR, but go into a PR on its own.